### PR TITLE
Fix shift selection for points

### PR DESCRIPTION
**Pull request by mstabrin**
# Description

Fixing a weird behavior with the fixpoint when shift is pressed during selection drag and points are already selected.
This happened, because the selection drag does not use centering of the data, but the newly created function did not 

## Type of change

- [X] Bug-fix (non-breaking change which fixes an issue)

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by mstabrin** [comment]
@Carreau @alisterburt @kevinyamauchi I found an issue with the Points layer when holding down shift to make a dragbox selection after PR https://github.com/napari/napari/pull/4999 

This PR will fix the issue :)

**Review by alisterburt — COMMENTED** [review]
Thanks for finding the bug @mstabrin and proposing a fix so quickly!

It's not immediately obvious to me why we wouldn't always center the drag start on the selection, could you explain some of the logic here? edit: when not modifying the selection, that is

**Comment by mstabrin** [comment]
From what I understand: It seems that we have four scenarios for the select event:
1. Click
  - All points are deselected, except the one below the cursor.
2. Drag
  - If there is a selection and the cursor is above a points, the points are dragged; if there is no selection a drag box to select the points within the box is shown.
3. Click with modifiers
  - The point below the cursor is added to the selection if not selected, it is deselected when already selected.
4. Drag with modifiers
  - A drag box to select the points within the box is shown. Selected box within the box are deselected, deselected ones are selected.

For scenarios 2 and 4 the drag box should start at the current cursor position (no center offset).
But scenarios 2 and 4 and dragging of points, we want to get the new shift relative to the old mouse position which by itself is relative to the old center, i.e., we want to keep the offset identical.

Therefore, for the drag box we do not want an offset, and for the movement of points we do want it :)

**Comment by alisterburt** [comment]
Thank you for the detailed explanation @mstabrin - I don't find the code here particularly easy to follow but I now understand it and it is a definite improvement over what we have. 

Could you add a simple test which verifies the behaviour you broke is fixed and would fail before this PR? Will be useful to catch in the future - I also added #5030 to track tidying this up :)

**Comment by mstabrin** [comment]
I added a "simple" test... However, the testing scenarios are not simple at all :D
For the shapes layer, I think the implementation is very inconsistent with how the Points layer behaves in the selection mode.

I tried the same tests like the Point layer, but I get very inconsistent results. When there is a rework going on, maybe the selection behavior can be unified somehow :S

**Comment by liu-ziyang** [comment]
I tried with both version on shift selection and then dragged, and I don't quite follow the behavior change. The current version of selection/deselection, shift selection/deselection, and dragging is working as expected. By any chance, can you please add a short video on the undesirable behavior being fixed? Thanks!

**Comment by psobolewskiPhD** [comment]
Playing with shift-drag, for me the selection box starts at the corner of the canvas to my mouse point, rather than starting at my mouse point. On live this doesn't happen.
Also, dragging selected points is much more performant on live, I see ~25% lower CPU usage (~40-80% vs. 300-400%).
Edit: when i checkout #4999 I get the reduced performance and big jump in CPU when moving points, vs. live, so that's not related to this PR.
Edit2: false alarm it goes back further than #4999

**Comment by mstabrin** [comment]
@potating-potato This issue was introduced by myself a few days ago in https://github.com/napari/napari/pull/4999.
As @psobolewskiPhD stated, it looks like this:
shift_drag_broken-2022-09-08_08 51 15

@psobolewskiPhD I also had the feeling that the performance decreased a bit, do you think it is due to #4999 ? Or do you mean you checked out the main before the merge.

**Comment by psobolewskiPhD** [comment]
@mstabrin I just checked a commit before #4999 and the performance reduction vs. live is present there too. I rarely use points, so I never noticed in previous usage of PRs and main. I will try to git bisect to find the commit.

**Comment by alisterburt** [comment]
@psobolewskiPhD you're a star! ⭐ I will add an issue to track

@potating-potato here is a video where the bug is quite pronounced

https://user-images.githubusercontent.com/7307488/189059943-e912b56d-51e7-47f2-99de-61500485992c.mp4

**Comment by psobolewskiPhD** [comment]
So the performance bug seems tied to my dev env, which diverged from my live env.
Even installing live into the env doesn't resolve, so it's something with a dependency I guess.
Edit: and it's not tied to points: pan-zoom on image is also less performant and CPU hungry. 👀 
Edit2: found the issue in case anyone is curious. The performant env has
```
libblas                   3.9.0           16_osxarm64_accelerate    conda-forge
libcblas                  3.9.0           16_osxarm64_accelerate    conda-forge
```
and the non-performant has:
```
libblas                   3.9.0           15_osxarm64_openblas    conda-forge
libcblas                  3.9.0           15_osxarm64_openblas    conda-forge
```
Making a fresh env with accelerate blas vs. openblas makes the issue clear.
Pretty crazy.

File: napari/layers/points/_tests/test_points_mouse_bindings.py [hunk 1]
```diff
@@ -681,3 +681,194 @@ def test_selecting_no_points_with_drag_3d(create_known_points_layer_3d):
...
+        ({1, 2}, False, []),
+        (set(), False, ['Shift']),
+        ({0}, False, ['Shift']),
+        ({0, 1, 2}, False, ['Shift']),
+        ({1, 2}, False, ['Shift']),
+    ],
+)
+def test_drag_start_selection(
+    create_known_points_layer_2d, pre_selection, on_point, modifier
+):
```
Comment (alisterburt) [inline napari/layers/points/_tests/test_points_mouse_bindings.py:709]:
```suggestion
def test_drag_start_selection(
    create_known_points_layer_2d, pre_selection, on_point, modifier
):
"""Check layer drag start and drag box behave as expected."""
```

File: napari/layers/shapes/_tests/test_shapes_mouse_bindings.py [hunk 2]
```diff
@@ -714,3 +714,169 @@ def test_all_modes_covered(attr):
...
+@pytest.mark.parametrize(
+    'pre_selection,on_point,modifier',
+    [
+        (set(), True, []),
+        ({1}, True, []),
+    ],
+)
+def test_drag_start_selection(
+    create_known_shapes_layer, Event, pre_selection, on_point, modifier
+):
```
Comment (alisterburt) [inline napari/layers/shapes/_tests/test_shapes_mouse_bindings.py:728]:
```suggestion
def test_drag_start_selection(
    create_known_shapes_layer, Event, pre_selection, on_point, modifier
):
"""Check layer drag start and drag box behave as expected."""
```

**Review by alisterburt — APPROVED** [review]
wow! Those tests are definitely complex, you likely know this machinery better than most right now 😄 

This PR fixes the bug and #4999 is not the source of the observed performance decrease so I'm approving here

**Comment by alisterburt** [comment]
if anyone from core-devs sees this feel free to merge once CI is passing

**Pull request merged**